### DISCUSSION TITLE
Trouble with syncing compute & render pipelines

### DISCUSSION THREAD

**Discussion by vini-fda** [gfx-rs/wgpu#3882] [Q&A]
For brief context: I'm trying to make a [simulation of the heat equation using wgpu](https://github.com/vini-fda/heat-wgpu), where every step is calculated using a compute shader in a double-buffered scheme. The result is then rendered in real-time.

Initially what I wanted was for the render and compute steps to be performed in separate threads, but that seems to be impossible/too difficult in the current state of wgpu. But now my implementation is single-threaded (again) and is performed within a single event loop sequentially: one compute step, then a render step every time there's an `Event::RedrawRequested`.  I do not perform polling on either steps to wait for the execution of the command buffers. The surface is configured with `PresentMode::Fifo` for VSync.

To give an overall idea, this is what is currently implemented:

- Two textures, A and B, which have `TEXTURE_BINDING` and `STORAGE_BINDING`  usages enabled. A is initialized with some noise.
- A compute step which has "forward" and "backward" texture binding groups. In "forward" mode, it reads from A and writes the results to B. The "backward" mode, as the name implies, does the opposite.
- A render step which also has "forward" and "backward" texture binding groups. In "forward" mode, it reads from A, while in "backward" mode, it reads from B. In both situations, the output is rendered to the surface texture (i.e. is shown in the screen)
- A variable `iteration` is incremented after every compute step, and it defines the direction in which both steps are performed (forward/backward)

Here's a diagram which explains it more succintly:

![image](https://github.com/gfx-rs/wgpu/assets/20410205/5d1e90c9-47ab-4923-bd19-98b1d7e3ea22)

The current issue is that many of the compute steps are being "dropped" between render steps, i.e. their command buffers are enqueued and submitted but are not being executed before their respective render step, which creates a situation in which most of the GPU work is just rendering, and not computing. This leads me to believe that I am not using the wgpu API the way it was intended. Is there a better way to do this?

For reference, here's a Firefox visualization of a profiling trace using [wgpu-profiler](https://github.com/Wumpf/wgpu-profiler):

![image](https://github.com/gfx-rs/wgpu/assets/20410205/1349d781-0a4d-412a-988f-a3807ccd825b)

**Comment by cwfitzgerald** [comment]
From what you described, there is nothing wrong with what you're doing. I would double check with something like renderdoc (using a custom capture: `device.start_capture()` `device.stop_capture()`) to see what's actually getting submitted. We don't do any kind of optimization on command buffers, so everything you submit should run.

**Reply by vini-fda** [reply]
I've never used renderdoc. Also, I'm on Wayland and it seems that it doesn't work properly with it. I'm using an NVIDIA GPU, though. Are NVIDIA profiling tools (e.g. NSight Graphics) OK for this purpose?

Edit: also, I'm not sure how to use the device capture. Am I supposed to start/stop it only once or at every compute/render step?

Edit 2: I've captured some traces with NSight Graphics (but I haven't added the custom captures that you suggested):

With VSync, the rendering barely gives opportunity for the compute kernel:

![image](https://github.com/gfx-rs/wgpu/assets/20410205/fb26bb4f-4ecb-4357-bbe1-32cf6b0f246f)

Without VSync, there seems to be proper syncing between computing and rendering:

![image](https://github.com/gfx-rs/wgpu/assets/20410205/58df03e8-1ec8-4960-90b5-edb1d1158500)
